File: algos/algos-0.0.1/ts/src/problem_solving/leet_code/group_anagrams/group_anagrams.ts
export function groupAnagrams(input: string[]): string[][] {
  const result: string[][] = []
  const store = new Map<string, string[]>()

  for (const word of input) {
    const sorted = word.split("").sort().join("")
    if (!store.has(sorted)) {
      store.set(sorted, [word])
    } else {
      store.get(sorted)!.push(word)
    }
  }

  for (const [_, value] of store) {
    result.push(value)
  }
  return result
}

File: algos/algos_depricated/go/problem_solving/leet_code/group_anagrams/group_anagrams.go
package group_anagrams

import "sort"

func groupAnagrams(list []string) [][]string {
	var result [][]string
	store := make(map[string][]string)
	for _, v := range list {
		sorted := sortString(v)
		if _, ok := store[sorted]; !ok {
			store[sorted] = []string{v}
		} else {
			store[sorted] = append(store[sorted], v)
		}
	}
	for _, v := range store {
		result = append(result, v)
	}
	return result
}

func sortString(s string) string {
	r := []rune(s)
	sort.Slice(r, func(i, j int) bool { return r[i] < r[j] })
	return string(r)
}
